cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java

```diff
@@ -117,7 +147,7 @@ public DynamicDatasetCache(SystemDatasetInstantiator instantiator,
     // The internal of the dataset instantiate may switch the context class loader to a different one when necessary
     ClassLoader currentClassLoader = ClassLoaders.setContextClassLoader(getClass().getClassLoader());
     try {
-      return getDataset(new DatasetCacheKey(name, dsArguments), bypass);
+      return getDataset(new DatasetCacheKey(name, dsArguments, accessType), bypass);
```

Comment:
We now cache entries twice if a dataset is used for read and write? What are the implications of that?